### ISSUE TITLE
Upward compressor: Incorrect implementation of "Boost" control

### ISSUE THREAD

**Issue by dstra2** [sadko4u/lsp-plugins#181]
### Problem

The "Boost" control of upward compressor behaves in an undesirable manner. An upward compressor should have a control which limits the maximum amplification level in order to keep the noise floor low during quiet parts of a track as well as to prevent over-compression. Unfortunately, in the current implementation the amplification limit is affected by 3 different controls (Boost, Ratio, Threshold) instead of only one, as shown in the following pictures, which makes it very difficult to adjust. Furthermore, it is not possible to set small amplification limits (of only a few dB).

Picture 1: Starting point with the following settings: Boost = -60dB, Threshold = -12dB, Ratio = 4:1 --> resulting amplification limit = +36 dB
![Pic_1](https://user-images.githubusercontent.com/24780672/107847902-b074fb80-6def-11eb-8793-b96d556bb368.png)

Picture 2: Change **Ratio** to 2:1 --> resulting amplification limit = +24dB
![Pic_2](https://user-images.githubusercontent.com/24780672/107847909-c4206200-6def-11eb-9cea-9d4071c42066.png)

Picture 3: Reset the controls to the values of picture 1, then change **Threshold** to -24dB --> resulting amplification limit = +27dB
![Pic_3](https://user-images.githubusercontent.com/24780672/107847922-ddc1a980-6def-11eb-92e0-ff3cff6581fe.png)

### Solution

It would be desirable
- to have only one knob (Boost) affecting and controlling the amplification limit
- to have the amplification limit range starting at 0dB (for very subtle interventions)
- to have the knob renamed to "Boost Limit", "Amp Limit" or "Gain Limit"

(It could also make sense to switch the positions of the Makeup and Boost knobs in the GUI since in the signal flow the amplification limit comes before Makeup.)

The following pictures (of Dynamic Processor) show the desirable behavior, i.e. changing Ratio or Threshold should have no effect on the amplification limit:

Picture 4: Starting point: Amp Limit = +24dB, Threshold = -12dB, Ratio = 3:1
![Pic_4](https://user-images.githubusercontent.com/24780672/107847932-fb8f0e80-6def-11eb-8e04-b86f6fa0424a.png)

Picture 5: Change **Ratio** to 2:1 --> amplification limit remains at +24dB
![Pic_5](https://user-images.githubusercontent.com/24780672/107847934-02b61c80-6df0-11eb-89be-3668d1a0ea2b.png)

Picture 6: Reset the controls to the values of picture 1, then change **Threshold** to -24dB --> amplification limit remains at +24dB
![Pic_6](https://user-images.githubusercontent.com/24780672/107847936-06e23a00-6df0-11eb-99ad-35904debcc67.png)

Apart from the above-mentioned issue, the upward compressor, especially the multiband version, can do "real magic" on mixes. Thank you for this great tool!

**Comment by sadko4u** [comment]
Hello!

Maybe you're interpreting the 'Boost' knob not in the right way.
This knob was added as a solution for issue #71

According to the description:
```
Boost - this knob is related to upward compression mode only. It defines the threshold below which the constant amplification will be applied to the input signal. This prevents the compressor from applying infinite amplification to very quiet signals. 
```

So, the knob sets the low threshold of the compressed curve. Consider you're setting the compression rate and then cancel it at some low threshold. The 'Boost' name was selected to minitmize the space taken by the label in the UI. If just renaming will solve the problem, then I believe adding 'Threshold' or 'Thresh' to the description would be enough.

What do you think?

**Comment by sadko4u** [comment]
The 'Dynamic processor' works in a bit another manner: it draws curve across all defined points and then smooths the knees. That's why you get another behaviour for Dynamic processor.

**Comment by dstra2** [comment]
Hi

Thank you for the answer. I was not aware that the "Boost" knob was intended as a solution to a technical problem.

I looked at it from another point of view:
- A compressor is nothing else than an automated mixing engineer sitting in front of a mixing desk and moving up and down a volume fader (very quickly) according the certain rules (compressor settings).
- Depending on the audio-material on a given track, the mixing engineer will have to decide how loud the quiet parts of the track should be in the final mix. E.g. he (respectively the artist) has to decide whether the breathing sound of a singer on a vocal track should be hearable or not. If he wants a quieter breathing sound, he will decide to not raise the volume fader above e.g. the +6dB mark while the singer is breathing. If he wants a louder breathing sound, he will decide to raise the fader e.g. to the +12dB mark, but not higher because otherwise unwanted sounds will become audible (smacking lips, rustling clothes, noise floor, ...). In other words: The mixing engineer makes a decision about the maximum amplification level for that given track.
- Since the mixing engineer delegates the fader moving work to the compressor, he has to tell the compressor about his decision regarding the maximum amplification level. In a downward compressor he accomplishes this by adjusting the makeup gain accordingly. In an upward compressor there's the need for another control to limit the amplification level.

The actual implementation of the "Boost" control as a threshold of the input signal does not correspond to what the mixing engineer wants to tell the compressor and how he wants it to behave. He needs a control to set the maximum amplification level (fader position of his mixing desk which shall not be exceeded...).

Of course this solution also solves the technical problem of unlimited amplification according to issue #71.

**Comment by sadko4u** [comment]
To not make people confusing about 'Boost', I think I should rename it in the interface to something neutral.
Don't you think so? Will it be sufficient?

**Comment by dstra2** [comment]
Maybe the title of this issue and the focus on the "Boost" control was misleading. It is more about a missing functionality. So simply renaming the control will not solve the issue.

As I tried to explain by the example above, from a mixing perspective there's the need to limit the maximum amplification of the compressor to a certain value, which is determined by musical and not technical considerations. Not being able to set this limit strongly reduces the usability of the upward compressor, because as soon as higher compression ratios are used, the noise floor or unwanted sounds become too loud (especially on audio material with pauses, where the compressor raises the gain).

For an upward compressor it is absolutely essential to have a control to set the amplification limit.

(And now the link to the "Boost" control: The amplification limit control would make the "Boost" control obsolete, so it would make sense to replace the "Boost" control by the "Boost/Amp/Gain limit" control. While the range of "Boost" control is negative [-120dB ... -60dB], the range of "Boost/Amp/Gain limit" control should be positive [0dB ... 60dB].)

Please let me know if you need further explanation.

**Comment by sadko4u** [comment]
@dstra2 OK, it's clear what you wanted to say.

You've described the behaviour of the compressor with amplification limit. For example, we amplify silent sounds by 12 dB and loud sounds keep non-amplified. The reasonable question is: when we should cancel the amplification and what's the transition zone from amplified signal to non-amplified signal.
If we set bot start point (below which we always amplify signal by 12 dB) and end point (above each se amplify signal by 0 dB), then we come back to the Dynamic processor plugin which works in exact way. But if our compressor starts working in such mode, we lose both boost and ratio control, too

![image](https://user-images.githubusercontent.com/12834464/107990487-9dc41780-6fe5-11eb-8488-6921fb180560.png)
Here we can see that Boost threshold (green), Ratio (blue lines, 4:1), Attack threshold (magenta) and amplification (red) influence one to another.

Now the gain boost at silent sounds is computed by using these parameters. If we cancel the Boost parameter and introduce the maximum amplification level, then we also are required to cancel the ratio parameter because with 1:1 ratio there's dilemma about how the compressor should work: amplify the signal to max amplification level or keep it unchanged to the original.

Also, changing the 'Boost' option will break the compatibility with other user sessions, and that's not good.

I think we can go a bit different way: I can introduce another compressor mode (don't know how to name it at this moment) where 'Boost' will control the maximum boost at the silent signal, 'Ratio', as usual, will control the ratio, and threshold will control the overall threshold. But how to deal with 1:1 ratio... I currently don't know.

I also wanted once to povide multiband dynamic processor as it is in the feature request list - #93. Maybe waiting for this issue would be an alternative option.

So let me please know what option would be the best.

**Comment by v-fox** [comment]
So, is this request anything like "S-shaped" compressor idea I was proposing in relation to #71 and ALR function of the limiter: a combination of upward and downward modes simultaneously ? Although, that still likely better realized by multiband compressor + [multiband ALR-limiter](https://github.com/sadko4u/lsp-plugins/issues/140) combo.

I find that, if anything, what's confusing about upward mode in the compressor is that you can't see the whole used curve in the graph down to -144dB, including the cut-off point. And that "gain" indicator is "glitched" in multiband compressor: it's capped for 24dB regardless of the real value.

**Comment by dstra2** [comment]
@sadko4u

> If we set bot start point (below which we always amplify signal by 12 dB) and end point (above each se amplify signal by 0 dB), then we come back to the Dynamic processor plugin which works in exact way. But if our compressor starts working in such mode, we lose both boost and ratio control, too

There is a difference between the proposed behaviour of the upward compressor and the Dynamic Processor. In Dynamic Processor, the start and end points are defined by the coordinates of the points ("threshold" and "gain" parameters). Therefore, these two points determine the ratio. In the proposed upward compressor only the end point is defined by a coordinate (on the diagonal dark grey 0dB gain line). The start point is a calculated point. It is the intersection of two lines:
- the line defined by the end point and the ratio parameter and
- the line defined by the amplification limit parameter (parallel line to the diagonal dark grey 0dB gain line)

Therefore, the ratio control will not get lost. (By varying the ratio parameter the start point moves diagonally up- or downwards)

> Now the gain boost at silent sounds is computed by using these parameters. If we cancel the Boost parameter and introduce the maximum amplification level, then we also are required to cancel the ratio parameter because with 1:1 ratio there's dilemma about how the compressor should work: amplify the signal to max amplification level or keep it unchanged to the original.

I am not shure whether I have understood the problem.
![Draft Gain Limit](https://user-images.githubusercontent.com/24780672/108428630-24276600-723f-11eb-8d72-0635efccfe57.gif)
The picture above shows the 3 cases that have to be taken into consideration. S1 and S2 are the starting points of the gain transition, T is the endpoint.
- Case 1 shows a higher ratio setting, where the compression curve hits the gain limit.
- Case 2 shows a lower ratio setting, where the gain limit will have no effect.
- Case 3 shows the 1:1 ratio setting, where the compressor does not perform any work (and therefore, it would make sense to bypass the compression routine in order to reduce CPU load as soon as the user sets this ratio).

The range of the diagram has been limited on the left side to "almost -infinity". By that, I mean that an input signal of 0 (= -infinity dBFS) is excluded from the diagram in order to be able to calculate the coordinates of point S2 (0 should be treated as a special case). "Almost -infinity" is the dBFS value of the smallest non-zero input signal that can be represented by the used data type (float32?).

> I think we can go a bit different way: I can introduce another compressor mode

In my view, introducing an additional compressor mode is a viable option.
In that case, I would also suggest an improvement of the current upward compressor mode: Increase the range of the "Boost" control up to -0 dB in order to make much lower (indirect) gain limits possible also in the current upward compressor mode. As there could arise the problem of conflicting boost and attack thresholds, the program code will probably have to ensure that the boost threshold is always lower than the attack threshold.

> I also wanted once to povide multiband dynamic processor as it is in the feature request list - #93. Maybe waiting for this issue would be an alternative option.

In my opinion, Dynamic Processor is a very versatile tool for special purposes, but less convenient for everyday use. Setting up Dynamic Processor and adjusting it to the processed audio material takes much more time than a compressor with the usual controls. Therefore, in my view, (Multiband) Dynamic Processor is not an alternative.

**Comment by dstra2** [comment]
@v-fox 

> So, is this request anything like "S-shaped" compressor idea I was proposing in relation to #71 and ALR function of the limiter: a combination of upward and downward modes simultaneously ? Although, that still likely better realized by multiband compressor + multiband ALR-limiter combo.

No, this request is not about combining upward and downward compression (S-shaped compression curve), it is about improving the usability of the upward compressor. I agree with you that in most cases an upward compressor (preserving transients) and a limiter (affecting only the loudest transients) in series is the better solution than a combined upward/downward compressor.

**Comment by sadko4u** [comment]
Committed all changes to **github-issue-181** branch.

**Comment by sadko4u** [comment]
Merged into **devel** branch: 0ec9c1c50d0d23ed99c6e83a618ff0a7533c741d

**Comment by sadko4u** [comment]
Removed branch **github-issue-181**.

**Comment by sadko4u** [comment]
Available in 1.1.30
